chef/cookbooks/ceph/recipes/radosgw_civetweb.rb

```diff
@@ -0,0 +1,79 @@
+# into radosgw.rb; it only exists as a separate file because it used to be
+# radosgw_apache2.rb
+
+if node["ceph"]["radosgw"]["ssl"]["enabled"]
+  certfile      = node["ceph"]["radosgw"]["ssl"]["certfile"]
+  keyfile       = node["ceph"]["radosgw"]["ssl"]["keyfile"]
+  pemfile       = node["ceph"]["radosgw"]["ssl"]["pemfile"]
+  if  node["ceph"]["radosgw"]["ssl"]["generate_certs"]
+    package "openssl"
+    ruby_block "generate_certs for radosgw" do
+        block do
+          unless ::File.exists? certfile and ::File.exists? keyfile
+            require "fileutils"
+
+            Chef::Log.info("Generating SSL certificate for radosgw...")
+
+            [:certfile, :keyfile].each do |k|
+              dir = File.dirname(node[:ceph][:radosgw][:ssl][k])
+              FileUtils.mkdir_p(dir) unless File.exists?(dir)
+            end
+
+            # Generate private key
+            %x(openssl genrsa -out #{keyfile} 4096)
+            if $?.exitstatus != 0
+              message = "SSL private key generation failed"
+              Chef::Log.fatal(message)
+              raise message
+            end
+            FileUtils.chown "root", node[:ceph][:group], keyfile
+            FileUtils.chmod 0640, keyfile
+
+            # Generate certificate signing requests (CSR)
+            conf_dir = File.dirname certfile
+            ssl_csr_file = "#{conf_dir}/signing_key.csr"
+            ssl_subject = "\"/C=US/ST=Unset/L=Unset/O=Unset/CN=#{node[:fqdn]}\""
+            %x(openssl req -new -key #{keyfile} -out #{ssl_csr_file} -subj #{ssl_subject})
+            if $?.exitstatus != 0
+              message = "SSL certificate signed requests generation failed"
+              Chef::Log.fatal(message)
+              raise message
+            end
+
+            # Generate self-signed certificate with above CSR
+            %x(openssl x509 -req -days 3650 -in #{ssl_csr_file} -signkey #{keyfile} -out #{certfile})
+            if $?.exitstatus != 0
+              message = "SSL self-signed certificate generation failed"
+              Chef::Log.fatal(message)
+              raise message
+            end
+
+            File.delete ssl_csr_file  # Nobody should even try to use this
+          end # unless files exist
+      end # block
+    end # ruby_block
+  else # if generate_certs
+    unless ::File.exists? certfile
+      message = "Certificate \"#{certfile}\" is not present."
+      Chef::Log.fatal(message)
+      raise message
+    end
+    # we do not check for existence of keyfile, as the private key is allowed
+    # to be in the certfile
+  end # if generate_certs
+ 
+  # Have to merge the SSL key and certificate into a pemfile, because that's what
+  # civetweb expects.  Per the comment above though, it's possible the keyfile
+  # may not actually exist, hence the existence check below
+  ruby_block "Creating radosgw pemfile" do
+    block do
+      %x(cp #{certfile} #{pemfile})
+      %x(cat #{keyfile} >> #{pemfile}) if ::File.exists? keyfile
```

Comment:
Lint/DeprecatedClassMethods: `File.exists?` is deprecated in favor of `File.exist?`.<br>Style/CommandLiteral: Use backticks around command string. (https://github.com/bbatsov/ruby-style-guide#percent-x)